doc/reference/kernel/other/fatal.rst

```diff
@@ -0,0 +1,245 @@
+to provide context to the assertion. In addition, the kernel will print
+a text representation of the expression code that was evaluated, and the
+file and line number where the assertion can be found.
+
+For example:
+
+.. code-block:: c
+
+  __ASSERT(foo == 0xF0CACC1A, "Invalid value of foo, got 0x%x", foo);
+
+If at runtime ``foo`` had some unexpected value, the error produced may
+look like the following:
+
+.. code-block:: none
+
+	ASSERTION FAIL [foo == 0xF0CACC1A] @ ZEPHYR_BASE/tests/kernel/fatal/src/main.c:367
+		Invalid value of foo, got 0xdeadbeef
+	[00:00:00.000,000] <err> os: r0/a1:  0x00000004  r1/a2:  0x0000016f  r2/a3:  0x00000000
+	[00:00:00.000,000] <err> os: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x00000a6d
+	[00:00:00.000,000] <err> os:  xpsr:  0x61000000
+	[00:00:00.000,000] <err> os: Faulting instruction address (r15/pc): 0x00009fe4
+	[00:00:00.000,000] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic
+	[00:00:00.000,000] <err> os: Current thread: 0x20000414 (main)
+	[00:00:00.000,000] <err> os: Halting system
+
+__ASSERT_EVAL()
+---------------
+
+The __ASSERT_EVAL() macro can also be used inside kernel and application code,
+with special semantics for the evalation of its arguments.
+
+It makes use of the __ASSERT() macro, but has some extra flexibility. It
+allows the developer to specify different actions depending whether the
+``__ASSERT()`` macro is enabled or not.  This can be particularly useful to
+prevent the compiler from generating comments (errors, warnings or remarks)
+about variables that are only used with ``__ASSERT()`` being assigned a value,
+but otherwise unused when the ``__ASSERT()`` macro is disabled.
+
+Consider the following example:
+
+.. code-block:: c
+
+  int x;
+  x = foo();
+  __ASSERT(x != 0, "foo() returned zero!");
+
+If __ASSERT() is disabled, then 'x' is assigned a value, but never used.
+This type of situation can be resolved using the __ASSERT_EVAL() macro.
+
+.. code-block:: c
+
+  __ASSERT_EVAL ((void) foo(),
+  		 int x = foo(),
+                 x != 0,
+                 "foo() returned zero!");
+
+The first parameter tells ``__ASSERT_EVAL()`` what to do if ``__ASSERT()`` is
+disabled.  The second parameter tells ``__ASSERT_EVAL()`` what to do if
+``__ASSERT()`` is enabled.  The third and fourth parameters are the parameters
+it passes to ``__ASSERT()``.
+
+__ASSERT_NO_MSG()
+-----------------
+
+The __ASSERT_NO_MSG() macro can be used to perform an assertion that reports
+the failed test and its location, but lacks additional debugging information
+provided to assist the user in diagnosing the problem; its use is
+discouraged.
+
+Build Assertions
+================
+
+Zephyr provides two macros for performing build-time assertion checks.
+These are evaluated completely at compile-time, and are always checked.
+
+BUILD_ASSERT_MSG()
+------------------
+
+This has the same semantics as C++'s ``static_assert``. If the evaluation
```

Comment:
`static_assert` is also present in C11 so "C/C++'s `static_assert`" might be appropriate, or "C's `_Static_assert` or C++'s `static_assert`" to reference the keyword form`.